### DISCUSSION TITLE
Java 11 project, unable to migrate from 0.9.1 to 0.16.0

### DISCUSSION THREAD

**Discussion by rchapin** [open-telemetry/opentelemetry-java#2878] [Q&A]
I have a Java 11 maven project in which I wrote a custom MetricExporter against the  io.opentelemetry.sdk.metrics.export.MetricExporter interface.

Previously, the dependencies in my pom were as follows 

    <dependency>
      <groupId>io.opentelemetry</groupId>
      <artifactId>opentelemetry-sdk</artifactId>
      <version>${opentelemetry.version}</version>
    </dependency>

    <dependency>
      <groupId>io.opentelemetry</groupId>
      <artifactId>opentelemetry-api</artifactId>
      <version>${opentelemetry.version}</version>
    </dependency>

    <dependency>
      <groupId>io.opentelemetry</groupId>
      <artifactId>opentelemetry-context-prop</artifactId>
      <version>${opentelemetry.version}</version>
    </dependency>

My module-info.java entries were

    requires transitive io.opentelemetry.api;
    requires transitive io.opentelemetry.sdk.common;
    requires transitive io.opentelemetry.sdk.metrics;

I updated the dependencies in the pom to

    <dependencyManagement>
      <dependencies>
        <dependency>
          <groupId>io.opentelemetry</groupId>
          <artifactId>opentelemetry-bom</artifactId>
          <version>${opentelemetry.version}</version>
          <type>pom</type>
          <scope>import</scope>
        </dependency>
      </dependencies>
    </dependencyManagement>
   
    <dependency>
      <groupId>io.opentelemetry</groupId>
      <artifactId>opentelemetry-sdk</artifactId>
      <version>${opentelemetry.version}</version>
    </dependency>

    <dependency>
      <groupId>io.opentelemetry</groupId>
      <artifactId>opentelemetry-api</artifactId>
      <version>${opentelemetry.version}</version>
    </dependency>

After attempting to upgrade to 0.16.0 I can no longer resolve any of imports from the ``` io.opentelemetry.sdk.metrics``` package or any of it's child packages.  I am unable to add the aforementioned package to my list of required modules and as a result cannot compile the program.

The list of modules output by the following command is identical between version 0.9.1 and 0.16.0.

```
$ jar --file=./opentelemetry-sdk-0.16.0.jar --describe-module
No module descriptor found. Derived automatic module.

io.opentelemetry.sdk@0.16.0 automatic
requires java.base mandated
contains io.opentelemetry.sdk
```
Any help would be greatly appreciated.

**Comment by anuraaga** [comment]
Hi @rchapin - sorry for the troubles. We've made many breaking changes since then, including separating metrics out to a separate artifact since it's not stable yet. You probably need to add the opentelemetry-sdk-metrics artifact to your build to get it back at the least.

To be honest, we'll probably continue to make significant breaks for a few months so I'd recommend a different metrics mechanism for now if it's hard to keep up with our releases.

**Reply by rchapin** [reply]
@anuraaga, thank you very much for getting back to me (and so quickly!).  Also, thank you for your candor about this project.

As I mentioned, I wrote a custom InfluxDB MetricExporter against 0.9.1 for a project that I am working on.  All of the unit and integration tests that I wrote for the exporter pass.  Once I started testing it in our dev environment against live data I noticed that, for some inexplicable reason it was dying right in the middle of the ``` public CompletableResultCode export(Collection<MetricData> metrics)``` where I am iterating over the Collection.

```
    @Override
    public CompletableResultCode export(Collection<MetricData> metrics) {
        String className;

        for (MetricData metricData : metrics) {
            /*
             * It is assumed that the name of the Metric is the name of the
             * measurement in which it will be recorded in InfluxDB.
             */
            String measurement = metricData.getName();

            Collection<Point> points = metricData.getPoints();
            for (Point point : points) {
                className = point.getClass().getSimpleName();
                LOGGER.log(
                    Level.FINER,
                    "Processing point: measurement={0}, className={1}, point={2}",
                    new Object[] { measurement, className, point});

--> Dies right here for some reason without any Exception

                org.influxdb.dto.Point influxDBPoint = null;
                try {
                    /*
                     * Dynamically instantiate the specific implementation
                     * PointProcessor class for the current point.
                     */
                    @SuppressWarnings("unchecked")
                    Class<PointProcessor> clazz = (Class<PointProcessor>) PointProcessors.get(className);
                    Constructor<PointProcessor> constructor = clazz.getDeclaredConstructor(
                        String.class, Point.class);
                    PointProcessor pointProcessor = constructor.newInstance(measurement, point);
                    influxDBPoint = pointProcessor.processPoint();
                    influxDB.write(influxDBPoint);
                } catch (Exception e) {
                    String stackTrace = ErrUtils.getStackTracetoString(e);
                    LOGGER.log(
                        Level.SEVERE,
                        "Unable to process metrics and send to InfluxDB for point={0}, " +
                        "influxDBPoint={1}, e={2}",
                        new Object[] {point, influxDBPoint, stackTrace});
                    return CompletableResultCode.ofFailure();
                }
            }
        }
        return CompletableResultCode.ofSuccess();
    }
```
As a test, I modified the method, wrapping the whole thing in a try/catch to see if there was some Exception that was being thrown by my code but did not see any evidence of it.

The only explanation that I could come up with was some concurrent modification problem in the OpenTelemetry library and that that Exception is being swallowed somewhere.

So, I wanted to see if a newer version would fix the issue.

Before I rip out OpenTelemetry and write my own stats collection implementation I would like to see if I can get the newest version running and if that fixes my issue.

The custom exporter that I wrote is not that much code, so it isn't that difficult to maintain, I am just stuck right now trying to get the current module-less OpenTelemetry project to work with my Java 11 project and any help would be greatly appreciated.

**Reply by rchapin** [reply]
@anuraaga there does not seem to be a ```opentelemetry-sdk-metrics``` artifact defined in the BOM, nor can I find it in maven central.  Any other ideas on the dependencies that I might be missing?

**Reply by jkwatson** [reply]
The main bom artifact only has the stable packages in it. You'll need to add the metrics dependency directly, or use our alpha-bom dependency to get it to resolve.

**Reply by rchapin** [reply]
@jkwatson thanks.  I figured that was the case regarding the alpha artifacts.  The question is, what dependency do I need to add?  As I mentioned, there is no ```opentelemetry-sdk-metrics``` in maven central that I can find.

**Reply by jkwatson** [reply]
https://search.maven.org/artifact/io.opentelemetry/opentelemetry-sdk-metrics/0.17.0-alpha/jar

**Reply by rchapin** [reply]
@anuraaga , @jkwatson thanks for your help.  Very much appreciated.  At this point, I've decided to swap out OpenTelemetry for another implementation because it seems like it's moving a bit too fast and I need reliable stats for this application.

Further, and maybe this is something that I should write up a issue for, IMHO the API needs the ability to reset any of the Instruments (LongValueRecorder, LongUpDownCounter, etc.).

It is a requirement for the team that I work with that any and all stats are verified in integration tests.  The problem with the OpenTelemetry library is that I do not have a deterministic way to reset any of the Instruments that I instantiate and as a result some of them retain state between my tests.  Firing up a new JVM for each test is a non-option and as a result I had to go through enormous hand-stands to be able to run reliable integration tests for the Exporter that I wrote.

For any other team that writes integration tests that validate stats having the ability to reset them, again IMHO, is an absolute requirement for any stats framework.

If there is another place that would make sense for me to detail these requirements for a feature request, please let me know.  Further, I'd be open to actually doing that work and contributing to the project as well.

**Reply by jkwatson** [reply]
If you're interested in this, I recommend joining the opentelemetry specification metrics working groups, where work is ongoing to define metrics data models and APIs. Putting issues into this repository would probably be premature, until the specifications are complete.

**Reply by rchapin** [reply]
OK, thanks.  Once I get this project off my plate I will see if I have the cycles.  Thanks again for your help.

**Reply by anuraaga** [reply]
Thanks for the understanding @rchapin! Hoping we get to a more stable state soon so you can try it out again.

**Reply by rchapin** [reply]
@anuraaga no problem whatsoever. I'm hoping that I can get a chance to contribute to the project.  :)  I posted a question about the API feature that I mentioned into the metrics specs discussion at gitter; we will see what everyone has to say.

**Reply by jkwatson** [reply]
Just FYI, the entire project is in the process of moving off of gitter and onto github discussions and CNCF slack. You might get more traction by creating a Discussion in the opentelemetry-specification repository, or in the `#otel-specification` channel on CNCF slack. (It's honestly hard to know where people are going to see things, but gitter is the least likely, I think).
